dialectic-compiler/src/lib.rs

```diff
@@ -0,0 +1,909 @@
+use {
+    lazy_static::lazy_static,
+    proc_macro2::{Span, TokenStream},
+    quote::{quote, ToTokens},
+    std::{fmt, ops, rc::Rc},
+    syn::{spanned::Spanned, Error, Ident, Type},
+    thiserror::Error,
+    thunderdome::{Arena, Index},
+};
+
+pub mod parse;
+
+#[derive(Error, Debug, Clone)]
+pub enum CompileError {
+    #[error("label name `'{0}` shadows a label name that is already in scope")]
+    ShadowedLabel(String),
+    #[error("undeclared label `'{0}`")]
+    UndeclaredLabel(String),
+    #[error("cannot `continue` outside of a loop")]
+    ContinueOutsideLoop,
+    #[error("cannot `break` outside of a loop")]
+    BreakOutsideLoop,
+}
+
+#[derive(Debug, Clone)]
+pub struct SpannedCompileError {
+    span: Span,
+    kind: CompileError,
+}
+
+/// A shim for parsing the root level of a macro invocation, so
+/// that a session type may be written as a block w/o an extra
+/// layer of braces.
+#[derive(Debug, Clone)]
+pub struct Invocation {
+    pub syntax: SyntaxNode,
+}
+
+#[derive(Debug, Clone)]
+pub struct SyntaxNode {
+    pub expr: Syntax,
+    pub span: Span,
+}
+
+#[derive(Debug, Clone)]
+pub enum Syntax {
+    Recv(Type),
+    Send(Type),
+    Call(Box<SyntaxNode>),
+    Choose(Vec<SyntaxNode>),
+    Offer(Vec<SyntaxNode>),
+    Split(Box<SyntaxNode>, Box<SyntaxNode>),
+    Loop(Option<String>, Box<SyntaxNode>),
+    Break(Option<String>),
+    Continue(Option<String>),
+    Block(Vec<SyntaxNode>),
+    Type(Type),
+}
+
+#[derive(Debug, Clone)]
+struct CfgNode {
```

Comment:
And here's another location that could use a shared `Spanned` type.

Reply:
This one contains some additional info besides a span though, and it seems a bit redundant to have `Spanned<CfgNode>` where this ends up essentially being two wrappers.